pimd/pim_sock.c

```diff
@@ -232,63 +216,84 @@ int pim_socket_mcast(int protocol, struct in_addr ifaddr, struct interface *ifp,
 			  rcvbuf);
 
 	{
-		long flags;
-
-		flags = fcntl(fd, F_GETFL, 0);
-		if (flags < 0) {
-			zlog_warn(
-				"Could not get fcntl(F_GETFL,O_NONBLOCK) on socket fd=%d: errno=%d: %s",
-				fd, errno, safe_strerror(errno));
-			close(fd);
-			return PIM_SOCK_ERR_NONBLOCK_GETFL;
-		}
-
-		if (fcntl(fd, F_SETFL, flags | O_NONBLOCK)) {
-			zlog_warn(
-				"Could not set fcntl(F_SETFL,O_NONBLOCK) on socket fd=%d: errno=%d: %s",
-				fd, errno, safe_strerror(errno));
-			close(fd);
-			return PIM_SOCK_ERR_NONBLOCK_SETFL;
-		}
+		set_nonblocking(fd);
 	}
 
+#if PIM_IPV == 4
 	/* Set Tx socket DSCP byte */
 	if (setsockopt_ipv4_tos(fd, IPTOS_PREC_INTERNETCONTROL)) {
 		zlog_warn("can't set sockopt IP_TOS to PIM/IGMP socket %d: %s",
 			  fd, safe_strerror(errno));
 	}
+#else
+		if (setsockopt_ipv6_tclass(fd, IPTOS_PREC_INTERNETCONTROL)) {
+			zlog_warn(
+				"can't set sockopt IP_TOS to PIMv6 socket %d: %s",
+				fd, safe_strerror(errno));
+		}
+#endif
 
-	return fd;
+	return 0;
 }
 
-int pim_socket_join(int fd, struct in_addr group, struct in_addr ifaddr,
-		    ifindex_t ifindex)
+int pim_socket_mcast(int protocol, pim_addr ifaddr, struct interface *ifp,
+		     uint8_t loop)
 {
+	int fd;
 	int ret;
 
-#ifdef HAVE_STRUCT_IP_MREQN_IMR_IFINDEX
-	struct ip_mreqn opt;
+	fd = pim_socket_raw(protocol);
+	if (fd < 0) {
+		zlog_warn("Could not create multicast socket: errno=%d: %s",
+			  errno, safe_strerror(errno));
+		return PIM_SOCK_ERR_SOCKET;
+	}
+
+#ifdef SO_BINDTODEVICE
+
+	ret = pim_socket_bind(fd, ifp);
+	if (ret) {
+		close(fd);
+		zlog_warn("Could not set fd: %d for interface: %s to device",
+			  fd, ifp->name);
+		return PIM_SOCK_ERR_BIND;
+	}
 #else
-	struct ip_mreq opt;
+/* XXX: use IP_PKTINFO / IP_RECVIF to emulate behaviour?  Or change to
+ * only use 1 socket for all interfaces? */
 #endif
+	ret = pim_setsockopt(protocol, fd, ifp, loop);
+	if (ret) {
+		close(fd);
+		zlog_warn("pim_setsockopt failed for interface: %s to device ",
+			  ifp->name);
+	}
+	return fd;
+}
 
-	opt.imr_multiaddr = group;
+int pim_socket_join(int fd, pim_addr group, pim_addr ifaddr, ifindex_t ifindex)
+{
+	int ret;
 
-#ifdef HAVE_STRUCT_IP_MREQN_IMR_IFINDEX
-	opt.imr_address = ifaddr;
-	opt.imr_ifindex = ifindex;
+#if PIM_IPV == 4
+	ret = setsockopt_ipv4_multicast(fd, IP_ADD_MEMBERSHIP, ifaddr,
+					group.s_addr, ifindex);
 #else
-	opt.imr_interface = ifaddr;
+	struct ipv6_mreq opt;
+
+	memcpy(&opt.ipv6mr_multiaddr, &group, 16);
+
+	opt.ipv6mr_interface = ifindex;
+
+	ret = setsockopt(fd, IPPROTO_IPV6, IPV6_JOIN_GROUP, &opt, sizeof(opt));
 #endif
 
-	ret = setsockopt(fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, &opt, sizeof(opt));
 	if (ret) {
-		char group_str[INET_ADDRSTRLEN];
-		char ifaddr_str[INET_ADDRSTRLEN];
-		if (!inet_ntop(AF_INET, &group, group_str, sizeof(group_str)))
+		char group_str[PIM_ADDRSTRLEN];
+		char ifaddr_str[PIM_ADDRSTRLEN];
+		if (!inet_ntop(PIM_AF, &group, group_str, sizeof(group_str)))
```

Comment:
There is no need to use inet_ntop, you can directly use %pPAs for pim_addr types.

Reply:
Ack